Guard logging calls to avoid unnecessary work at call site

stephentoub opened the pull request:
(Ignore the first commit. It is a copy of the commit from https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2161, which this PR then builds on.)

Many LogHelper.Log calls are doing work at the call site, such as allocating params arrays, boxing structs, formatting strings, and so on, even when the data will be thrown away because logging (or logging for that verbosity level) isn't enabled.  This adds an IsEnabled method to LogHelper and uses it at any call site where there's such work to be avoided.

Fixes https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2163

jennyf19 commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -159,7 +160,9 @@ public virtual bool CanReadToken(string token)
             if (token.Length > MaximumTokenSizeInBytes)
             {
-                LogHelper.LogInformation(TokenLogMessages.IDX10209, LogHelper.MarkAsNonPII(token.Length), LogHelper.MarkAsNonPII(MaximumTokenSizeInBytes));
+                if (LogHelper.IsEnabled(EventLogLevel.Informational))
```
nice catch

jennyf19 reviewed (approved):
This is great. We should the same in our other library as well, I think we have a similar issue there.

keegan-caruso commented:
@stephentoub merge conflict from #2161

stephentoub commented:
> merge conflict from https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2161

Yup, thanks, fixed.  I'd submitted this with the commit from that one, just deleted that first commit here with a rebase.

jennyf19 commented:
@stephentoub looks like there are still merge conflicts?

stephentoub commented:
> looks like there are still merge conflicts?

Not "still", but "again" :)  I'll fix them again.

jennyf19 commented:
> > looks like there are still merge conflicts?
> 
> Not "still", but "again" :) I'll fix them again.

Hard to keep up with you!

stephentoub commented:
Conflicts resolved.

jennyf19 commented:
@keegan-caruso  and @brentschmaltz the tests via the GitHub action is not working properly, or there is a test failure not bubbling up nicely. We will have to test this one locally for now.

keegan-caruso commented:
@jennyf19 I don't understand, when I pull this one locally the test run hangs

jennyf19 commented:
> @jennyf19 I don't understand, when I pull this one locally the test run hangs

That makes sense...But that test doesn't normally hang?

BrennanConroy commented on src/Microsoft.IdentityModel.Xml/DsigSerializer.cs:
```diff
@@ -110,27 +111,29 @@ public virtual KeyInfo ReadKeyInfo(XmlReader reader)
                         // Skip the element since it is not one of  <RetrievalMethod>, <X509Data>, <KeyValue>
-                        LogHelper.LogWarning(LogMessages.IDX30300, reader.ReadOuterXml());
+                        if (LogHelper.IsEnabled(EventLogLevel.Warning))
+                            LogHelper.LogWarning(LogMessages.IDX30300, reader.ReadOuterXml());
```
Does this change behavior? `reader.ReadOuterXml()` isn't called unless logging is on now.

Should it call `reader.Skip()` in an `else`?

stephentoub replied:
Ugh, yes, it does, and should, or at the very least it should store the result into a local outside of the if. Good spot.

stephentoub committed:
Guard logging calls to avoid unnecessary work at call site

Many LogHelper.Log calls are doing work at the call site, such as allocating params arrays, boxing structs, formatting strings, and so on, even when the data will be thrown away because logging (or logging for that verbosity level) isn't enabled.  This adds an IsEnabled method to LogHelper and uses it at any call site where there's such work to be avoided.
